src/agent.py

```diff
@@ -113,33 +104,24 @@ def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
         if not container.can_connect():
             event.defer()
             return
-        if not event.unit or not all(
-            event.relation.data[event.unit].get(required_data)
-            for required_data in ("executors", "labels", "name")
-        ):
+        # The relation is joined, it cannot be None, hence the type casting.
+        agent_relation_meta = typing.cast(
+            typing.Mapping[str, AgentMeta], self.state.agent_relation_meta
+        )
+        # The event unit cannot be None.
+        agent_meta = agent_relation_meta[typing.cast(ops.Unit, event.unit).name]
```

Comment:
@benhoyt interesting case, if the unit cannot be None should the type hint be changed?

Reply:
Yeah, there's almost certainly room for improvement there. The doc comment says about `RelationEvent.unit`:

```
     This will be ``None`` if the relation event was triggered as an
    :class:`Application <model.Application>`-level event.
```

I don't actually know what that means!

I also have a comment on `RelationEvent.app` that says, "# TODO(benhoyt): I *think* app should never be None, but confirm and update type".

It's quite possible the more specific event types, like `RelationJoinedEvent`, can have narrower attribute types. Feel free to open a `canonical/operator` issue to look into improving this.

Reply:
Can do!